util/HalideTraceUtils.h

```diff
@@ -0,0 +1,91 @@
+#ifndef HALIDE_TRACE_UTILS_H
+#define HALIDE_TRACE_UTILS_H
+
+#include "HalideRuntime.h"
+#include <stdio.h>
+
+namespace Halide {
+namespace Internal {
+
+void bad_type_error(halide_type_t type);
+
+// Simple conversion.
+template<typename T>
```

Comment:
Could be done with less casting via halide_scalar_value_t:

```
template<typename T>
T value_as(halide_type_t type, const void* value) {
    const halide_scalar_value_t *v = (const halide_scalar_value_t *)value;
    switch (type.code) {
    case halide_type_int:
        switch (type.bits) {
        case 8:
            return (T) v->i8;
        case 16:
            return (T) v->i16;
        case 32:
            return (T) v->i32;
        case 64:
            return (T) v->i64;
        default:
            bad_type_error(type);
        }
        break;
    case halide_type_uint:
        switch (type.bits) {
        case 1:
            return (T) v->b;
        case 8:
            return (T) v->u8;
        case 16:
            return (T) v->u16;
        case 32:
            return (T) v->u32;
        case 64:
            return (T) v->u64;
        default:
            bad_type_error(type);
        }
        break;
    case halide_type_float:
        switch (type.bits) {
        case 32:
            return (T) v->f32;
        case 64:
            return (T) v->f64;
        default:
            bad_type_error(type);
        }
        break;
    default:
        bad_type_error(type);
    }
    return (T) 0;
}
```

Reply:
Thanks, copy-pasted